Sarus CI

haampie opened the pull request:
This is still a bit of a WIP. 

For security reasons I would like to avoid running generated shell commands outside of the container, so I'm thinking about adding a bot that will comment on pull requests to remind you to update the gitlab pipeline definition whenever a new test is added or flags are changed.

rasolca commented on docker/build-env/Dockerfile:
```diff
@@ -0,0 +1,137 @@
+ARG BOOST_PATCH=0
+ARG BOOST_PATH=/usr/local/boost
+
+RUN wget -q https://dl.bintray.com/boostorg/release/${BOOST_MAJOR}.${BOOST_MINOR}.${BOOST_PATCH}/source/boost_${BOOST_MAJOR}_${BOOST_MINOR}_${BOOST_PATCH}.tar.gz -O boost.tar.gz && \
+    tar -xzf boost.tar.gz && \
+    cd boost_${BOOST_MAJOR}_${BOOST_MINOR}_${BOOST_PATCH} && \
+    ./bootstrap.sh --prefix=$BOOST_PATH && \
+    ./b2 -j$(nproc) && \
+    ./b2 install && \
+    rm -rf /root/boost.tar.gz /root/boost_${BOOST_MAJOR}_${BOOST_MINOR}_${BOOST_PATCH}
```
I think we need only Boost headers.
@albestro Isn't HPX needing only them?

haampie replied:
ok, that's good to know. At any rate, the dependencies are not recompiled every time, but cached inside of the build environment docker image

albestro replied:
We depend on HPX 1.4.0
https://github.com/eth-cscs/DLA-Future/blob/4d92d14ba7e22bbcaa3475013626d9842f9572cc/CMakeLists.txt#L63

which [release notes](https://stellar-group.github.io/hpx/docs/sphinx/tags/1.4.0/html/releases/whats_new_1_4_0.html) states (I'm not really sure how to interpret it)

>We no longer require compiled Boost dependencies in certain configurations. This requires at least Boost 1.70, compiling on x86 with GCC 9, clang (libc++) 9, or VS2019 in C++17 mode.

In this issues you can track the status of the Boost dependencies in HPX https://github.com/STEllAR-GROUP/hpx/issues/3440

hkaiser replied:
HPX does not depend on any compiled Boost libraries anymore if 
- you are using Boost >= V1.70 (Boost.System is header only)
- you are using C++17 and a compiler that supports `std::filesystem`
- you are not interested in building and running the HPX inspect tool (depends on Boost.Regex)

We also reduce HPX's dependency on other Boost libraries as well, this however will take more time. The mentioned tickets documents our progress.

rasolca commented on docker/build-env/Dockerfile:
```diff
@@ -0,0 +1,137 @@
+FROM nvidia/cuda:10.1-devel-ubuntu18.04
```
This image is good for testing on GPUs (nothing to test yet, but we will need this later...).
Please add a non CUDA image for testing on multicore.

haampie replied:
I was thinking to build one big build environment container with all prebuilt dependencies in it, and then do a multistage build where a fresh ubuntu image is extended and binaries are copied over from the previous stage. If you decide not to build for GPUs, then you simply don't copy the binaries over. This process is automated (see below)

rasolca replied:
My concern is that libraries may be compiled in a different way if CUDA is found or not (For example HPX).
If possible I would like to run the tests for a multicore build in an environment without traces of CUDA.

rasolca commented on docker/deploy/Dockerfile:
```diff
@@ -0,0 +1,73 @@
+      make -j$(nproc) && \
+      # Let's just bundle the libs to make the docker image small
+      # We have to copy a couple of MKL libs by hand
+      /root/libtree/libtree --strip --chrpath \
+        -d /root/DLA-Future.bundle \
+        $TEST_BINARIES \
+        ${MKL_LIB}/libmkl_tbb_thread.so \
+        ${MKL_LIB}/libmkl_intel_thread.so \
+        ${MKL_LIB}/libmkl_sequential.so \
+        ${MKL_LIB}/libmkl_def.so \
+        ${MKL_LIB}/libmkl_vml_def.so \
+        ${MKL_LIB}/libmkl_vml_cmpt.so \
+        ${MKL_LIB}/libmkl_mc.so \
+        ${MKL_LIB}/libmkl_vml_mc.so \
+        ${MKL_LIB}/libmkl_mc3.so \
+        ${MKL_LIB}/libmkl_vml_mc3.so \
+        ${MKL_LIB}/libmkl_avx.so \
+        ${MKL_LIB}/libmkl_vml_avx.so \
+        ${MKL_LIB}/libmkl_avx2.so \
+        ${MKL_LIB}/libmkl_vml_avx2.so \
+        ${MKL_LIB}/libmkl_avx512_mic.so \
+        ${MKL_LIB}/libmkl_vml_avx512_mic.so \
+        ${MKL_LIB}/libmkl_avx512.so \
+        ${MKL_LIB}/libmkl_vml_avx512.so \
+        ${MKL_LIB}/libmkl_core.so && \
```
Can you please explain what this step does and why only a part of the mkl's libraries is present?

haampie replied:
The build environment docker image is huge (5.94GB uncompressed atm), so what I'm doing to reduce it in size is to copy over only essential binaries for the tests using https://github.com/haampie/libtree. It figures out like `ldd` what the dependent shared libs are, and copies them to a single folder. MKL however uses `dlopen` to load microarchitecture specific libraries, so you have to copy those implicit dependencies yourself by listing them like that. The resulting image is "only" 711MB uncompressed (mostly due to MKL...).

Currently it is somewhat important to keep images small, because we cannot build them on Daint yet and therefore build and store them temporarily elsewhere. That means there is overhead of creating tarballs, but also storage can be an issue at some point (currently Gitlab is used to store images, which has a 10GB limit per repo for their Docker registry)

rasolca commented on docker/deploy/Dockerfile:
```diff
@@ -0,0 +1,73 @@
+
+FROM $BUILD_ENV as builder
+
+ENV MKL_LIB=/opt/intel/compilers_and_libraries/linux/mkl/lib/intel64
```
I recommend to
`source /opt/intel/compilers_and_libraries/linux/mkl/bin/mklvars.sh intel64`
to define MKL's environment and use
`MKL_LIB=$MKLROOT/lib/intel64`

haampie replied:
Ah yes, I think I have tried it previously, but then I had not set the shell to bash for it to work yet

rasolca commented on docker/deploy/Dockerfile:
```diff
@@ -0,0 +1,73 @@
+
+RUN mkdir DLA-Future-build && cd DLA-Future-build && \
+    CC=/usr/local/mpich/bin/mpicc CXX=/usr/local/mpich/bin/mpicxx cmake /DLA-Future \
+      -DMKL_ROOT=/opt/intel/compilers_and_libraries/linux/mkl \
```
if the MKL environment is set, this is not needed.

haampie commented:
> Unfortunately this approach doesn't work with PRs. Imagine that two PRs are added and each one adds a different test. How can you modify the pipeline to run all the tests of both?

True, in that case let me just copy the generated `~/run.sh` shell script from the container and execute that on daint.

haampie commented:
bors ping

haampie commented:
bors try

haampie commented:
@rasolca, could you comment `bors try` on this PR? If should build + run already :). See https://bors.tech/documentation/ for reference

rasolca commented:
bors try

rasolca commented:
> @rasolca, could you comment `bors try` on this PR? If should build + run already :). See https://bors.tech/documentation/ for reference

I gave you write access to the repository too. Is it enough or you need more permissions to run bors?

haampie commented:
Weird, seems like networking issues in Kubernetes or so. Now trying with `--network=host`.

bors try

haampie commented:
It compiles, but it hits that hpx issue it seems... Let me run it one more time to see if caching of binaries works properly.

bors try

haampie commented:
Now with the same config as in jenkins:

bors try

haampie commented:
bors try

haampie commented:
typo in the pipelines, again

bors try

haampie commented:
bors try

haampie commented:
bors try

albestro commented on docker/deploy/Dockerfile:
```diff
@@ -0,0 +1,75 @@
+# Make it easy to call our binaries.
+ENV PATH="/root/DLA-Future.bundle/usr/bin:$PATH"
+
+RUN echo "/root/DLA-Future.bundle/usr/lib/" > /etc/ld.so.conf.d/cosma.conf && ldconfig
```
```suggestion
RUN echo "/root/DLA-Future.bundle/usr/lib/" > /etc/ld.so.conf.d/dlaf.conf && ldconfig
```

albestro commented:
Another thing I've just realized.

We have a `check-format` step, in particular pay attention to:
- [No trailing whitespaces](https://github.com/eth-cscs/DLA-Future/blob/master/misc/coding.md#horizontal-spacing)
- [Exactly one newline before EOF](https://github.com/eth-cscs/DLA-Future/blob/master/misc/coding.md#vertical-spacing)

haampie commented:
bors try

haampie commented:
`check-format` can also be a github action

haampie commented:
I've added a debug build here as well that uses OpenBLAS to keep things reasonably small and sprinkles around `-fsanitize -Og -g` in HPX and dlaf. One issue currently is that the amazing BLAS detection in blaspp compiles and executes a binary, which requires certain Linux capabilities it cannot have. Let's see how that goes in Sarus then :)

bors try

haampie commented:
bors ping

haampie commented:
bors try

haampie commented:
bors try-

haampie commented:
bors try

haampie commented:
Oh... github refuses to run bors whenever new github actions are part of the PR. Let me make that a separate PR then.

haampie commented:
bors try

haampie commented:
bors try-

haampie commented:
bors try

haampie commented:
bors try

haampie commented:
@rasolca, see the above... tests succeed in debug mode but _not_ in release mode :( and sanitizer tools are not working with hpx at the moment due to https://github.com/STEllAR-GROUP/hpx/issues/3824.

rasolca commented:
@haampie Warning: I don't consider the debug tests passed. For example see the last lines of this output:
```
+ srun -n 6 -c 5 --jobid=22080048 unbuffer sarus run -t --mpi registry.gitlab.com/cscs-ci/eth-cscs/dla-future/deploy-debug:8fc880d2a72003c19936f413ba70d0a7939050ac test_util_matrix --hpx:use-process-mask
Running main() from gtest_mpihpx_main.cpp
[==========] Running 12 tests from 4 test suites.
[----------] Global test environment set-up.
Running main() from gtest_mpihpx_main.cpp
Running main() from gtest_mpihpx_main.cpp
Running main() from gtest_mpihpx_main.cpp
Running main() from gtest_mpihpx_main.cpp
Running main() from gtest_mpihpx_main.cpp
[----------] 3 tests from MatrixUtilsTest/0, where TypeParam = float
[ RUN      ] MatrixUtilsTest/0.Set
[       OK ] MatrixUtilsTest/0.Set (3 ms)
[ RUN      ] MatrixUtilsTest/0.SetRandom
[       OK ] MatrixUtilsTest/0.SetRandom (3 ms)
[ RUN      ] MatrixUtilsTest/0.SetRandomHermitianPositiveDefinite
[       OK ] MatrixUtilsTest/0.SetRandomHermitianPositiveDefinite (4 ms)
[----------] 3 tests from MatrixUtilsTest/0 (10 ms total)
[----------] 3 tests from MatrixUtilsTest/1, where TypeParam = double
[ RUN      ] MatrixUtilsTest/1.Set
[       OK ] MatrixUtilsTest/1.Set (2 ms)
[ RUN      ] MatrixUtilsTest/1.SetRandom
[       OK ] MatrixUtilsTest/1.SetRandom (2 ms)
[ RUN      ] MatrixUtilsTest/1.SetRandomHermitianPositiveDefinite
[       OK ] MatrixUtilsTest/1.SetRandomHermitianPositiveDefinite (3 ms)
[----------] 3 tests from MatrixUtilsTest/1 (7 ms total)
[----------] 3 tests from MatrixUtilsTest/2, where TypeParam = std::complex<float>
[ RUN      ] MatrixUtilsTest/2.Set
[       OK ] MatrixUtilsTest/2.Set (2 ms)
[ RUN      ] MatrixUtilsTest/2.SetRandom
[       OK ] MatrixUtilsTest/2.SetRandom (9 ms)
[ RUN      ] MatrixUtilsTest/2.SetRandomHermitianPositiveDefinite
[       OK ] MatrixUtilsTest/2.SetRandomHermitianPositiveDefinite (3 ms)
[----------] 3 tests from MatrixUtilsTest/2 (15 ms total)
[----------] 3 tests from MatrixUtilsTest/3, where TypeParam = std::complex<double>
[ RUN      ] MatrixUtilsTest/3.Set
[       OK ] MatrixUtilsTest/3.Set (1 ms)
[ RUN      ] MatrixUtilsTest/3.SetRandom
[       OK ] MatrixUtilsTest/3.SetRandom (3 ms)
[ RUN      ] MatrixUtilsTest/3.SetRandomHermitianPositiveDefinite
[       OK ] MatrixUtilsTest/3.SetRandomHermitianPositiveDefinite (3 ms)
[----------] 3 tests from MatrixUtilsTest/3 (7 ms total)
[----------] Global test environment tear-down
[==========] 12 tests from 4 test suites ran. (42 ms total)
[  PASSED  ] 12 tests.
srun: First task exited 30s ago
srun: step:22080048.9 tasks 0,4: running
srun: step:22080048.9 tasks 1-3,5: exited
srun: Terminating job step 22080048.9
srun: Job step aborted: Waiting up to 92 seconds for job step to finish.
```

haampie commented:
True... no error code from srun apparently.

Here's an example of a test that succeeds in debug mode https://gitlab.com/cscs-ci/eth-cscs/DLA-Future/-/jobs/518036134#L293 but fails in release mode https://gitlab.com/cscs-ci/eth-cscs/DLA-Future/-/jobs/518036135#L293

haampie commented:
bors try

haampie commented:
It seems the issue is not container-specific. When I copy things from the container like this, I get the same errors:

```bash
container=`docker run -d -it registry.gitlab.com/cscs-ci/eth-cscs/dla-future/deploy-release:ddb82d57b46e6b66f0bcf8a258e8c465ccffd443 /bin/bash`
docker cp $container:/root/DLA-Future.bundle .
docker stop $container
scp -r DLA-Future.bundle daint:...
ssh daint
salloc -N 1 -C mc -p cscsci
LD_LIBRARY_PATH=/opt/cray/pe/mpt/7.7.9/gni/mpich-gnu-abi/7.1/lib/ srun -n 6 -c 5 ./DLA-Future.bundle/usr/bin/test_matrix -Ihpx.stacks.small_size=0x80000 --hpx:use-process-mask
```

rasolca commented:
I created a vm with ubuntu18.04 and installed all the dependencies as done in the docker image.
Then I copied the files from the container and executed them. Everything works fine.
I suspect that there is some incompatibility problems when MPI is replaced on daint.
```
rasolca@dlaf_experiments> container=`sudo docker run -d -it registry.gitlab.com/cscs-ci/eth-cscs/dla-future/deploy-release:ddb82d57b46e6b66f0bcf8a258e8c465ccffd443 /bin/bash`
rasolca@dlaf_experiments> sudo docker cp $container:/root/DLA-Future.bundle .
rasolca@dlaf_experiments> sudo docker stop $container
c16299d79fa14fc2d3f0107b3033b53c3c69ad2b5fc4164b0bf2f043bf25d360

rasolca@dlaf_experiments> cd DLA-Future.bundle/usr/bin
rasolca@dlaf_experiments> /usr/local/mpich/bin/mpirun -n 6 ./test_matrix --hpx:use-process-mask
Running main() from gtest_mpihpx_main.cpp
Running main() from gtest_mpihpx_main.cpp
Running main() from gtest_mpihpx_main.cpp
[==========] Running 107 tests from 10 test suites.
[----------] Global test environment set-up.
Running main() from gtest_mpihpx_main.cpp
Running main() from gtest_mpihpx_main.cpp
Running main() from gtest_mpihpx_main.cpp
[----------] 9 tests from MatrixLocalTest/0, where TypeParam = float
[ RUN      ] MatrixLocalTest/0.StaticAPI
[       OK ] MatrixLocalTest/0.StaticAPI (0 ms)
[ RUN      ] MatrixLocalTest/0.StaticAPIConst
[       OK ] MatrixLocalTest/0.StaticAPIConst (0 ms)
[ RUN      ] MatrixLocalTest/0.Constructor
[       OK ] MatrixLocalTest/0.Constructor (1 ms)
.........
[ RUN      ] MatrixExceptionPropagation.ReadDoesNotPropagateInRWAccess
[       OK ] MatrixExceptionPropagation.ReadDoesNotPropagateInRWAccess (0 ms)
[ RUN      ] MatrixExceptionPropagation.ReadDoesNotPropagateInReadAccess
[       OK ] MatrixExceptionPropagation.ReadDoesNotPropagateInReadAccess (0 ms)
[----------] 4 tests from MatrixExceptionPropagation (1018 ms total)

[----------] Global test environment tear-down
[==========] 107 tests from 10 test suites ran. (10923 ms total)
[  PASSED  ] 107 tests.
```

haampie commented:
Ok, let's try with the latest mpich.

bors try

haampie commented:
bors try

haampie commented:
bors ping

haampie commented:
bors try

haampie commented:
bors try-

haampie commented:
bors try

haampie commented:
without the small stack size flag

bors try

haampie commented:
I've cleaned this up to a single commit and incorporated your feedback @rasolca. Also the AddressSanitizer builds are removed. I got it to work with clang though, but it seems to give quite some spurious errors as well. We could look at it later...

bors try

haampie commented:
bors try

haampie commented:
```
 [222229.659262047] [nid00025-14157] [Runtime] [WARN] Mounting /dev/pts without the gid=<gid of tty group> option, because no tty gid was found. Some programs, e.g. sshd, might run into errors because of this.
 Failed to retrieve username for uid=25014 from "/apps/daint/system/opt/sarus/1.1.0/etc/passwd"
```

🤷‍♂️ 

bors try

rasolca commented on docker/deploy-debug.Dockerfile:
```diff
@@ -0,0 +1,54 @@
+      -DDLAF_WITH_TEST=ON \
+      -DDLAF_BUILD_MINIAPPS=OFF \
+      -DDLAF_MPI_PRESET=slurm \
+      -DMPIEXEC_NUMCORES=32 \
```
Why 32?
```suggestion
      -DMPIEXEC_NUMCORES=36 \
```

rasolca commented on docker/deploy-release.Dockerfile:
```diff
@@ -0,0 +1,61 @@
+      -DDLAF_WITH_TEST=ON \
+      -DDLAF_BUILD_MINIAPPS=OFF \
+      -DDLAF_MPI_PRESET=slurm \
+      -DMPIEXEC_NUMCORES=32 \
```
```suggestion
      -DMPIEXEC_NUMCORES=36 \
```

rasolca commented on cmake/DLAF_AddTest.cmake:
```diff
@@ -171,6 +171,8 @@ function(DLAF_addTest test_target_name)
+    set(_TEST_LABEL MPI)
   # ----- Classic test
   else()
     set(_TEST_COMMAND ${test_target_name})
```
What about setting a label here as well? (E.g. NOMPI)

haampie replied:
I've added this, but later refactored the shell script not to make use of it. It still there since it cannot hurt to have labels...

rasolca commented on docker/deploy-debug.Dockerfile:
```diff
@@ -0,0 +1,54 @@
+      -DDLAF_WITH_CUDA=OFF \
+      -DDLAF_WITH_MKL=OFF \
+      -DDLAF_WITH_TEST=ON \
+      -DDLAF_BUILD_MINIAPPS=OFF \
```
I would build miniapps too, to test if they still compile.

rasolca commented on docker/deploy-release.Dockerfile:
```diff
@@ -0,0 +1,61 @@
+      -DDLAF_WITH_CUDA=OFF \
+      -DDLAF_WITH_MKL=ON \
+      -DDLAF_WITH_TEST=ON \
+      -DDLAF_BUILD_MINIAPPS=OFF \
```
See above.

rasolca commented on ci/sarusify.sh:
```diff
@@ -0,0 +1,13 @@
+TEST_BINARIES=`ctest -L MPI -V -N | grep -oP "(?<=\")(/root/DLA-Future-build.*?)(?=\")"`
+
+# Copy all test commands over to a single shell script
+SARUS_TEST_COMMANDS=$(ctest -L MPI -V -N | sed -n -e 's/^.*Test command: //p' | sed 's|/usr/bin/srun|srun|g')
```
Non MPI tests are not executed...
Can the command `srun -n 1 -c 36 sarus run ctest -L NOMPI` be added?

haampie commented:
bors try

haampie commented:
Ugh. I promise this will be the last run :|

bors try

rasolca commented:
bors try

haampie commented:
how about another

bors try

rasolca commented:
bors try

rasolca commented:
bors try

rasolca commented:
bors try

haampie commented:
So, two comments:

1. curiously in release mode this happens:
    
    ```
    srun: First task exited 30s ago
    srun: step:22153127.9 task 2: running
    srun: step:22153127.9 tasks 0-1,3: exited
    srun: Terminating job step 22153127.9
    srun: Job step aborted: Waiting up to 92 seconds for job step to finish.
    ```

2. And the addr2line stuff works as well, which is convenient. From the logs in debug mode you see the url of the docker image + a backtrace like this:
    
    ```
    Backtrace:
    test_util_math(+0x2cb8f)[0x555555580b8f]
    test_util_math(+0x2eef8)[0x555555582ef8]
    test_util_math(+0x54e30)[0x5555555a8e30]
    test_util_math(+0x5b106)[0x5555555af106]
    test_util_math(+0x4d037)[0x5555555a1037]
    ```
    
    And then with a single command you can locally get the line numbers where the segfault has happened:
    
    ```
    $ docker run registry.gitlab.com/cscs-ci/eth-cscs/dla-future/deploy-debug:23db8e6dd25f79c796fd44eb05f8f94e10e11ea6 addr2line -e test_util_math 0x2cb8f 0x2eef8 0x54e30 0x5b106 0x4d037
    Unable to find image 'registry.gitlab.com/cscs-ci/eth-cscs/dla-future/deploy-debug:23db8e6dd25f79c796fd44eb05f8f94e10e11ea6' locally
    23db8e6dd25f79c796fd44eb05f8f94e10e11ea6: Pulling from cscs-ci/eth-cscs/dla-future/deploy-debug
    5bed26d33875: Already exists 
    f11b29a9c730: Already exists 
    930bda195c84: Already exists 
    78bf9a5ad49e: Already exists 
    81e3830ca117: Pull complete 
    a68345418c3f: Pull complete 
    d8a345682de1: Pull complete 
    4f4fb700ef54: Pull complete 
    Digest: sha256:fb53b302ec685b341db3a806b3a24b6cf5b795b666429925b6e02802c5398ade
    Status: Downloaded newer image for registry.gitlab.com/cscs-ci/eth-cscs/dla-future/deploy-debug:23db8e6dd25f79c796fd44eb05f8f94e10e11ea6
    /DLA-Future/test/unit/test_util_math.cpp:64
    /DLA-Future/test/unit/test_util_math.cpp:74
    /root/DLA-Future-build/_deps/googletest-src/googletest/src/gtest.cc:2435 (discriminator 4)
    /root/DLA-Future-build/_deps/googletest-src/googletest/src/gtest.cc:2469
    /root/DLA-Future-build/_deps/googletest-src/googletest/src/gtest.cc:2508
    ```

rasolca commented:
I retried just the release tests and the slurm problem disapperared.

albestro commented on docker/build-env-release.Dockerfile:
```diff
@@ -0,0 +1,137 @@
+    cmake .. \
+      -DBOOST_ROOT=$BOOST_PATH \
+      -DHWLOC_ROOT=$HWLOC_PATH \
+      -DTCMALLOC_ROOT=$GPERFTOOLS_PATH \
```
Have you checked if you get the expected configuration? I would check if `HPX_WITH_MALLOC` has to be set manually or it is automatically detected.

(In case, see also `build-env-debug.Dockerfile`)

haampie replied:
Seems like it is picked up, from the logs it seems `libhwloc.so.15.2.0` is linked against :)

albestro commented on docker/build-env-release.Dockerfile:
```diff
@@ -0,0 +1,137 @@
+    make install -j$(nproc) && \
+    rm -rf /root/mpich-${MPICH_VERSION}.tar.gz /root/mpich-${MPICH_VERSION}
+
+# Install MKL
+ARG MKL_VERSION=2020.0-088
+RUN wget -qO - https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS-2019.PUB 2>/dev/null | apt-key add - && \
+    apt-add-repository 'deb https://apt.repos.intel.com/mkl all main' && \
+    apt-get install -y -qq --no-install-recommends intel-mkl-64bit-${MKL_VERSION} && \
+    rm -rf /var/lib/apt/lists/* && \
+    echo "/opt/intel/lib/intel64\n/opt/intel/compilers_and_libraries/linux/mkl/lib/intel64" >> /etc/ld.so.conf.d/intel.conf && \
+    ldconfig
```
I see that in release you use MKL, while in debug OpenBLAS. Is it different on purpose?

haampie replied:
I was thinking that you would have debug symbols for all dependencies now, even BLAS. With MKL I don't think you get them (or they would not be of any use anyways because it's closed source).

albestro replied:
Good point! I don't know what @rasolca thinks about this, but if one side it is very nice to have two different dependencies built and tested (MKL and OpenBLAS), at the same time I'm wondering if it wouldn't be better to build the same configuration in debug and release.

rasolca replied:
At some point we should define the matrix of the configurations we want to test (maybe a smaller matrix for PRs and a bigger one for weekly(?) tests), but it is not responsibility of @haampie.

albestro commented on docker/deploy-debug.Dockerfile:
```diff
@@ -0,0 +1,51 @@
+# Make it easy to call our binaries.
+ENV PATH="/root/DLA-Future.bundle/usr/bin:$PATH"
+
+# Automatically print stacktraces on segfault
+ENV LD_PRELOAD=/lib/x86_64-linux-gnu/libSegFault.so
```
Is it worth having it just for debug? I know without symbols the stacktrace would not be "detailed", I'm just asking if you already tested/considered this option.

rasolca commented:
bors try

rasolca commented:
bors try

rasolca commented:
bors try

rasolca commented:
bors try

rasolca commented:
Hoping that this is the last
bors try

rasolca commented:
bors try

rasolca commented:
bors try
